### ISSUE TITLE
Variables from inner mixin override variables in parents

### ISSUE THREAD

**Issue by newkind** [leafo/scssphp#440]
Looks like the `master` version of compiler stopped properly recognizing mixin multiple values for ie. Bourbon `flex()`

This code which is proper and was working fine before :

`@include flex(X Y);`

now excludes Y and only considers X. 

Sample code :

`@include flex(0 100%);`

After compilation using `master` branch :

`flex: 0;`

Should be :
```
flex: 0 100%;
```
The only solution right now is to stop using this mixin and break it into two separate rules.

**Comment by w00fz** [comment]
This is the sassmeister gist https://gist.github.com/w00fz/6ee4744601ef81992ce03cd79398b743
You can see the output is not consistent with the SCSS compiler.

**Comment by mahagr** [comment]
This works:

```
#foo { @include flex(0 100%); }
@media only all and (max-width: 47.938rem) { #foo { @include flex(0 100%); } }
```

But if you have nested includes, it won't work:

```
@include breakpoint(mobile-only) { #foo { @include flex(0 100%); } }
```

**Comment by w00fz** [comment]
After further testing, it looks like it's the nested `@include` causing the issue. Consider the following example (using bourbon for simplicity):

```sass
// Load Bourbon
@import "vendor/bourbon/bourbon";

#foo {
    @include flex(0 100%);
}

@include hidpi(1.5) {
    #foo {
        @include flex(0 100%);
    }
}
```

This is the output after compilation:

```css
#foo {
  -webkit-box-flex: 0;
  -moz-box-flex: 0;
  box-flex: 0;
  -webkit-flex: 0 100%;
  -moz-flex: 0 100%;
  -ms-flex: 0 100%;
  flex: 0 100%;
}
@media only screen and (-webkit-min-device-pixel-ratio: 1.5), only screen and (min--moz-device-pixel-ratio: 1.5), only screen and (-o-min-device-pixel-ratio: 1.5/1), only screen and (min-resolution: 144dpi), only screen and (min-resolution: 1.5dppx) {
  #foo {
    -webkit-box-flex: 0;
    -moz-box-flex: 0;
    box-flex: 0;
    -webkit-flex: 0;
    -moz-flex: 0;
    -ms-flex: 0;
    flex: 0;
  }
}
```

So as you can see, the second output is not rendering as expected.

**Comment by mahagr** [comment]
Alright, I found the cause, this is minimal code to reproduce the issue without any external dependencies:

```scss
@mixin prefixer($property, $value) {
  #{$property}: $value;
}

@mixin outer() {
  @content;
}

@mixin inner($value) {
  value: $value;
  @include prefixer(foo, 'xxx');
  value: $value;
}

#foo {
  @include inner(0 100%);
}

@include outer() {
  #foo {
    @include inner(0 100%);
  }
}
```

Result looks like:
```css
/* line 17, test.scss */
#foo {
  value: 0 100%;
  foo: 'xxx';
  value: 0 100%;
}
/* line 22, test.scss */
#foo {
  value: 0 100%;
  foo: 'xxx';
  value: 'xxx';
}
```

**Comment by mahagr** [comment]
The issue seems to be that getStoreEnv() always points to the same object, which makes the inner arguments to override the original values in the outer store.

**Comment by mahagr** [comment]
@robocoder I think I found a fix for this bug. Basically what I'm doing is to change the storeEnv to the current one just after push and then restore it before pop.

```
            case Type::T_INCLUDE:
                // including a mixin
                list(, $name, $argValues, $content) = $child;

                $mixin = $this->get(self::$namespaces['mixin'] . $name, false);

                if (! $mixin) {
                    $this->throwError("Undefined mixin $name");
                    break;
                }

                $callingScope = $this->getStoreEnv();

                // push scope, apply args
                $this->pushEnv();
                $this->env->depth--;

                $storeEnv = $this->storeEnv;
                $this->storeEnv = $this->env;

                if (isset($content)) {
                    $content->scope = $callingScope;

                    $this->setRaw(self::$namespaces['special'] . 'content', $content, $this->env);
                }

                if (isset($mixin->args)) {
                    $this->applyArguments($mixin->args, $argValues);
                }

                $this->env->marker = 'mixin';
                $this->compileChildrenNoReturn($mixin->children, $out);

                $this->storeEnv = $storeEnv;

                $this->popEnv();
                break;
```

**Comment by mahagr** [comment]
OK, the fix seems to break something else, some variables are now missing...

**Comment by mahagr** [comment]
Ok, same is needed for function calls:

```
    protected function callScssFunction($name, $argValues, &$returnValue)
    {
        $func = $this->get(self::$namespaces['function'] . $name, false);

        if (! $func) {
            return false;
        }

        $this->pushEnv();

        $storeEnv = $this->storeEnv;
        $this->storeEnv = $this->env;

        // set the args
        if (isset($func->args)) {
            $this->applyArguments($func->args, $argValues);
        }

        // throw away lines and children
        $tmp = new OutputBlock;
        $tmp->lines    = [];
        $tmp->children = [];

        $this->env->marker = 'function';

        $ret = $this->compileChildren($func->children, $tmp);

        $this->storeEnv = $storeEnv;

        $this->popEnv();

        $returnValue = ! isset($ret) ? self::$defaultValue : $ret;

        return true;
    }
```

**Comment by mahagr** [comment]
Here is test case for the function calls:

```scss
@function contains($list, $values...) {
  @each $value in $values {
    @if type-of(index($list, $value)) != "number" {
      @return false;
    }
  }

  @return true;
}

@function mapping($items) {
  $src: ();
  $map: (
          one:   1 100,
          two:   2 200,
  );

  @each $key, $values in $map {
    @if contains($items, $key) {
      $value1: nth($values, 1);
      $value2: nth($values, 2);

      $src: append($src, $value1 $value2, comma);
    }
  }

  @return $src;
}

@mixin test() {
  test: mapping(one two);
}

test: mapping(one two);
@include test();
```

**Comment by robocoder** [comment]
Looks promising. I look forward to seeing your pull request.

**Comment by mahagr** [comment]
Fixes also #370

**Comment by mahagr** [comment]
Looks the same as #348

**Comment by robocoder** [comment]
@mahagr : 👍

**Comment by w00fz** [comment]
Nice Matias!

@robocoder, are you considering merging @mahagr and doing a release in the very near future?
Our product, Gantry 5, relies on your great compiler a lot and we get a lot of feedback regarding SCSS compilations issues (which aren't many!), this is also how we noticed the issue above.

We are currently in the process of doing a major release and we upgraded to the latest SCSS compiler and Bourbon, so far this has been the only issue we noticed or that has been reported.

We can always use a fork with the patch or if you just merge in, we could lock the dependency to that commit, but it's been a while since the last release and would be really great if we could just use a release version as dependency.

Thank you for your hard work.
